### PR TITLE
Stop requiring a pystub for register_fake by default

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124299
* #124200
* #124199
* #124134
* #124180
* #124092
* #124089
* #124071
* #124066
* #124065
* __->__ #124064
* #123937

Previously, if someone used `register_fake` to add a fake impl for an
operator defined in C++, we would require them to add a
`m.set_python_module(<module>)` call to C++. This was to avoid
situations where a user imported the C++ operator without importing the
fake impl.

This "breaks" open registration: there's no way to add a fake impl
outside of a repository that defines an operator, so we want to turn
this behavior off by default in open source.

Test Plan:
- existing tests

### CODE REVIEW

File: torch/_utils_internal.py [hunk 1]
```diff
@@ -164,3 +164,4 @@ def max_clock_rate():
 # USE_RTLD_GLOBAL_WITH_LIBTORCH controls whether __init__.py tries to load
 # _C.so with RTLD_GLOBAL during the call to dlopen.
 USE_RTLD_GLOBAL_WITH_LIBTORCH = False
+REQUIRE_PYTHON_REGISTRATION_STUB_FOR_CPP_OPS = False
```
Comment (albanD) [inline torch/_utils_internal.py:167]: Can you have a detailed comment like the above?

**Review by albanD — COMMENTED** [review]
Why do we want to preserve the current behavior?

File: torch/_utils_internal.py (hunk 1 above)
Comment (zou3519) [reply torch/_utils_internal.py]: done

File: torch/_library/utils.py [hunk 2]
```diff
@@ -189,3 +190,12 @@ def can_generate_trivial_fake_impl(op: torch._ops.OpOverload) -> bool:
...
     return True
+
+
+def requires_set_python_module() -> bool:
+    """If an op was defined in C++ and extended from Python using the
+    torch.library APIs, returns if we require that there have been a
+    m.set_python_module("mylib.ops") call from C++ that associates
+    the C++ op with a python module.
+    """
+    return getattr(_utils_internal, "REQUIRES_SET_PYTHON_MODULE", True)
```
Comment (albanD) [inline torch/_library/utils.py:201]: nit: I guess the default value is not really needed here since this option is always defined in the file below?

**Review by albanD — APPROVED** [review]
Thanks for the more detailed comment!

File: torch/_library/utils.py (hunk 2 above)
Comment (zou3519) [reply torch/_library/utils.py:201]: We need it to stage the change with internal

**Pull request closed without merging**